pytorch_lightning/core/lightning.py

```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
             optimizer: Current optimizer used in training_loop
             optimizer_idx: Current optimizer idx in training_loop
         """
+
+        # Iterate over all optimizer parameters to preserve their `requires_grad` information
+        # in case these are pre-defined during `configure_optimizers`
         param_requires_grad_state = {}
-        # make sure current optimizer is latest to be iterated over.
-        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
-        num_optimizers = len(optimizers) - 1
-        for opt_idx, opt in enumerate(optimizers):
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
```

Comment:
Can `param_requires_grad_state` be calculated once per training step instead of once per optimizer per training step?

Reply:
It is possible, but it will make use of `toggle_optimizer` slightly harder to properly use for new users.